frameworks/Haskell/servant/mysql-haskell/src/Main.hs

```diff
@@ -0,0 +1,256 @@
+
+------------------------------------------------------------------------------
+
+-- * Test 1: JSON serialization
+
+json :: Handler Aeson.Value
+json = return . Aeson.Object $ fromList [("message", "Hello, World!")]
+{-# INLINE json #-}
+
+
+-- * Test 2: Single database query
+
+decodeWorld :: DbRow -> Either Text World
+decodeWorld [] = Left "MarshalError: Expected 2 columns for World, found 0"
+decodeWorld (_:[]) = Left "MarshalError: Expected 2 columns for World, found 1"
+decodeWorld (c1:c2:_) = World <$> intValDec c1 <*> intValDec c2
+{-# INLINE decodeWorld #-}
+
+extractWorld :: Streams.InputStream DbRow -> IO (Either Text World)
+extractWorld rowsS = do
+  rows <- Streams.toList rowsS
+  return $ case rows of
+    [] -> Left "No rows found!"
+    (row:_) -> decodeWorld row
+
+singleDb :: DbPool -> GenIO -> Handler World
+singleDb pool gen = do
+  v <- liftIO $ uniformR (1, 10000) gen
+  r <- liftIO $ Pool.withResource pool $ \conn -> do
+    (_, rowsS) <- MySQL.query conn "SELECT * FROM World WHERE id = ?" [intValEnc v]
+    extractWorld rowsS
+  case r of
+    Left e -> throwError err500 { errBody = LBS.fromStrict $ TextEnc.encodeUtf8 e }
+    Right world -> return world
+{-# INLINE singleDb #-}
+
+-- * Test 3: Multiple database query
+
+multipleDb :: DbPool -> GenIO -> Maybe Count -> Handler [World]
+multipleDb pool gen mcount = do
+  res <- liftIO $ Pool.withResource pool $ \conn -> do
+    sId <- MySQL.prepareStmt conn "SELECT * FROM World WHERE id = ?"
+    res <- replicateM (getCount mcount) $ do
+      v <- uniformR (1, 10000) gen
+      (_, rowsS) <- MySQL.queryStmt conn sId [intValEnc v]
+      extractWorld rowsS
+    MySQL.closeStmt conn sId
+    return res
+  let (errs, oks) = partitionEithers res
+  case errs of
+    [] -> return oks
+    e -> throwError err500 { errBody = LBS.fromStrict . TextEnc.encodeUtf8 . Text.pack . show $ e }
+{-# INLINE multipleDb #-}
+
+
+-- * Test 4: Fortunes
+
+decodeFortune :: DbRow -> Either Text Fortune
+decodeFortune [] = Left "MarshalError: Expected 2 columns for Fortune, found 0"
+decodeFortune (_:[]) = Left "MarshalError: Expected 2 columns for Fortune, found 1"
+decodeFortune (c1:c2:_) = Fortune <$> intValDec c1 <*> textValDec c2
+{-# INLINE decodeFortune #-}
+
+selectFortunes :: MySQL.MySQLConn -> IO (Either [Text] [Fortune])
+selectFortunes conn = do
+  (_, rowsS) <- MySQL.query_ conn "SELECT * FROM Fortune"
+  rows <- Streams.toList rowsS
+  let eFortunes = fmap decodeFortune rows
+  let (err, oks) = partitionEithers eFortunes
+  return $ case err of
+    [] -> pure oks
+    _ -> Left err
+{-# INLINE selectFortunes #-}
+
+fortunes :: DbPool -> Handler (Lucid.Html ())
+fortunes pool = do
+  r <- liftIO $ Pool.withResource pool selectFortunes
+  case r of
+    Left e -> throwError err500 { errBody = LBS.fromStrict . TextEnc.encodeUtf8 . Text.pack . show $ e }
+    Right fs -> return $ do
+      let new = Fortune 0 "Additional fortune added at request time."
+      Lucid.doctypehtml_ $ do
```

Comment:
(`type-of-html`)[http://hackage.haskell.org/package/type-of-html] might be substantially faster.

Reply:
Already working on it in a separate branch ;)

Reply:
https://github.com/TechEmpower/FrameworkBenchmarks/pull/4553